### PR TITLE
bootloader: set proper LISTEN_PID in child process

### PR DESCRIPTION
**Pull request by zhsj**
if the program is actived by a systemd socket, then there are some
envs(LISTEN_PID, LISTEN_FDS) for it.

LISTEN_PID is set to the pid which is forked by systemd that is the
parent process of the bootloader.

This patch is trying to correct the env to the child pid of bootloader.
parent fds are duplicated to child process, so it's OK to modify
LISTEN_PID.

I test it with gunicorn which [supports](http://docs.gunicorn.org/en/stable/deploy.html#systemd) systemd socket.

Some [related code](https://github.com/benoitc/gunicorn/blob/c54426c1e4b26b30c5feee2cec4450656ae147a3/gunicorn/systemd.py#L36-L39) in gunicorn which detects the LISTEN_PID env.

### CODE REVIEW

**Comment by zhsj** [comment]
I polished this patch. But I don't know why the CI failed. The error in the log seems not related to this modification.

**Comment by zhsj** [comment]
Hi, what's the situation of this commit? anything I can do?

File: bootloader/src/pyi_utils.c [hunk 1]
```diff
@@ -740,6 +740,19 @@ pyi_utils_set_environment(const ARCHIVE_STATUS *status)
     return rc;
 }
 
+int
+set_systemd_env(pid_t pid)
+{
+    #define ULONG_STRING_SIZE (sizeof (unsigned long) * CHAR_BIT / 3 + 2)
+    const char * env_var = "LISTEN_PID";
+    char pid_str[ULONG_STRING_SIZE];
+    snprintf(pid_str, ULONG_STRING_SIZE, "%ld", (unsigned long)pid);
```
Comment (htgoebel) [inline bootloader/src/pyi_utils.c]: Move this into the `if`. formatting is only required if the variable need to be set.

File: bootloader/src/pyi_utils.c [hunk 2]
```diff
@@ -740,6 +740,19 @@ pyi_utils_set_environment(const ARCHIVE_STATUS *status)
     return rc;
 }
 
+int
+set_systemd_env(pid_t pid)
+{
+    #define ULONG_STRING_SIZE (sizeof (unsigned long) * CHAR_BIT / 3 + 2)
```
Comment (htgoebel) [inline bootloader/src/pyi_utils.c]: Looks weird. What *exactly* does it do?

File: bootloader/src/pyi_utils.c [hunk 3]
```diff
@@ -789,6 +802,7 @@ pyi_utils_create_child(const char *thisfile, const int argc, char *const argv[])
     /* Child code. */
     if (pid == 0) {
         /* Replace process by starting a new application. */
+        set_systemd_env(getpid());
```
Comment (htgoebel) [inline bootloader/src/pyi_utils.c]: I'd move the `getpid()` call into the `if` within `set_systemd_env`. The the resp. env-var is not set, nothing needs to be done at all.

File: bootloader/src/pyi_utils.c [hunk 4]
```diff
@@ -740,6 +740,19 @@ pyi_utils_set_environment(const ARCHIVE_STATUS *status)
     return rc;
 }
 
+int
+set_systemd_env(pid_t pid)
+{
```
Comment (htgoebel) [inline bootloader/src/pyi_utils.c]: Please add a comment, what this does and why. Basically this si what you've written in the commit message. E.g.

> if the program is actived by a systemd socket, LISTEN_PID ist set and needs to be changed to the child pid of the bootloader

**Comment by htgoebel** [comment]
@codewarrior0 Can you please review this. Thanks.

File: bootloader/src/pyi_utils.c (hunk 2 above)
Comment (zhsj) [inline bootloader/src/pyi_utils.c]: emm, I don't know what the pid_str size I should set, so this line roughly calculates the max decimal string length...Any better idea?

File: bootloader/src/pyi_utils.c [hunk 5]
```diff
@@ -740,6 +740,26 @@ pyi_utils_set_environment(const ARCHIVE_STATUS *status)
...
+/*
+ * If the program is actived by systemd socket, LISTEN_PID is set to parent
+ * pid of the bootloader and should be changed to child pid.
+ */
+int
+set_systemd_env()
+{
+    const char * env_var = "LISTEN_PID";
+    if(pyi_getenv(env_var) != NULL) {
+        #define ULONG_STRING_SIZE (sizeof (unsigned long) * CHAR_BIT / 3 + 2)
```
Comment (htgoebel) [inline bootloader/src/pyi_utils.c]: While still waiting for review by @codewarrior0: Where is this from? Please add a source. Thanks.

File: bootloader/src/pyi_utils.c [hunk 6]
```diff
@@ -740,6 +740,26 @@ pyi_utils_set_environment(const ARCHIVE_STATUS *status)
     return rc;
 }
 
+/*
+ * If the program is actived by systemd socket, LISTEN_PID is set to parent
+ * pid of the bootloader and should be changed to child pid.
+ */
```
Comment (htgoebel) [inline bootloader/src/pyi_utils.c]: Please put most parts of the commit message here, thus the reason for the code is easier to understand by just reading the code. Thanks.

Comment (zhsj) [inline bootloader/src/pyi_utils.c]: adjusted!

Comment (htgoebel) [inline bootloader/src/pyi_utils.c]: Thanks :-)

**Comment by htgoebel** [comment]
@codewarrior0 I'd appreciate your review of these view lines, master of the C-code. This will be the last change to the bootloader for the next release. After you reviewed, I can rebuild the bootloaders.

File: bootloader/src/pyi_utils.c [hunk 7]
```diff
@@ -859,6 +891,7 @@ pyi_utils_create_child(const char *thisfile, const int argc, char *const argv[])
     /* Child code. */
     if (pid == 0) {
         /* Replace process by starting a new application. */
+        set_systemd_env();
```
Comment (codewarrior0) [inline bootloader/src/pyi_utils.c]: If this function is going to return a result code, you should at least check it and print a warning message on failure.

Other than that, looks good to me!

Comment (zhsj) [inline bootloader/src/pyi_utils.c]: thanks! adjusted.

**Comment by htgoebel** [comment]
Thank you very much do both of you. Now I can prepare the bootloaders for the next release.

**Pull request merged**